File: factory-method.java
// Certainly! The Factory Method design pattern is used to define an interface for creating an object, but let subclasses alter the type of objects that will be created. Here's a complex example of the Factory Method pattern in Java:

// Copy
// Product interface
interface Car {
    void drive();
}

// Concrete products
class Sedan implements Car {
    @Override
    public void drive() {
        System.out.println("Driving a Sedan");
    }
}

class SUV implements Car {
    @Override
    public void drive() {
        System.out.println("Driving an SUV");
    }
}

// Creator interface
interface CarFactory {
    Car createCar();
}

// Concrete creators
class SedanFactory implements CarFactory {
    @Override
    public Car createCar() {
        return new Sedan();
    }
}

class SUVFactory implements CarFactory {
    @Override
    public Car createCar() {
        return new SUV();
    }
}

// Client code
public class FactoryMethodPatternExample {
    public static void main(String[] args) {
        CarFactory sedanFactory = new SedanFactory();
        Car sedan = sedanFactory.createCar();
        sedan.drive();

        CarFactory suvFactory = new SUVFactory();
        Car suv = suvFactory.createCar();
        suv.drive();
    }
}
In this example:

Car is the product interface that defines the common method drive.
Sedan and SUV are concrete product classes that implement the Car interface.
CarFactory is the creator interface that declares the factory method createCar.
SedanFactory and SUVFactory are concrete creator classes that implement the CarFactory interface and provide the specific implementation of the createCar method.
The client code (FactoryMethodPatternExample) demonstrates how different factories (creators) can be used to create specific types of cars (products). This allows for flexibility in creating different types of objects without changing the client code.

The Factory Method pattern is especially useful when you have a family of related classes, and the exact class to be used is determined at runtime. In this example, the client code can easily switch between creating Sedans and SUVs by using the appropriate factory without modifying the existing code.

File: factory-method.go
// Certainly! Here's an example of the Factory Method design pattern implemented in Go:

// Copy
package main

import "fmt"

// Product interface
type Car interface {
	Drive()
}

// Concrete products
type Sedan struct{}

func (s *Sedan) Drive() {
	fmt.Println("Driving a Sedan")
}

type SUV struct{}

func (s *SUV) Drive() {
	fmt.Println("Driving an SUV")
}

// Creator interface
type CarFactory interface {
	CreateCar() Car
}

// Concrete creators
type SedanFactory struct{}

func (sf *SedanFactory) CreateCar() Car {
	return &Sedan{}
}

type SUVFactory struct{}

func (suvf *SUVFactory) CreateCar() Car {
	return &SUV{}
}

// Client code
func main() {
	sedanFactory := &SedanFactory{}
	sedan := sedanFactory.CreateCar()
	sedan.Drive()

	suvFactory := &SUVFactory{}
	suv := suvFactory.CreateCar()
	suv.Drive()
}
In this example:

Car is the product interface that declares the common method Drive.
Sedan and SUV are concrete product structs that implement the Car interface.
CarFactory is the creator interface that declares the factory method CreateCar.
SedanFactory and SUVFactory are concrete creator structs that implement the CarFactory interface and provide the specific implementation of the CreateCar method.
The client code demonstrates how different factories (creators) can be used to create specific types of cars (products). The factories encapsulate the logic for creating objects, allowing the client code to be decoupled from the actual class instantiation.

This example illustrates how the Factory Method pattern can be applied in Go to create various types of cars without modifying the existing client code. The client can easily switch between creating Sedans and SUVs by using the appropriate factory.